### PR TITLE
Add global lock to avoid multiple goroutines using the mutexes map at the same time

### LINKED ISSUE #7: Crashed after a while with 'concurrent map read/write'
Ran for ages and then crashed.

```
fatal error: concurrent mapread and map write
```

```
/home/ops/specs-bot/src/specs-bot/main.go:104 +0x606
```

Full stacktrace:

```
INFO[2019-04-01T19:09:22.925829217Z] Received webhook
INFO[2019-04-01T19:09:22.957585078Z] Received webhook
fatal error: concurrent mapread and map write

goroutine 389 [running]:
runtime.throw(0x948b31, 0x21)
        /root/go/src/runtime/panic.go:616 +0x81 fp=0xc420505e68 sp=0xc420505e48 pc=0x434981
runtime.mapaccess2_fast64(0x89f3e0, 0xc4200992f0, 0x79a, 0x87d2c0, 0xc4205bc0a0)
        /root/go/src/runtime/hashmap_fast.go:141 +0x1aafp=0xc420505e90 sp=0xc420505e68 pc=0x41535a
mutex.Lock(0x79a)
        /home/ops/specs-bot/src/mutex/mutex.go:13 +0x47fp=0xc420505ed0 sp=0xc420505e90 pc=0x772867
hook.HandlePullRequestPayload(0xc4205100d8, 0x7, 0x79a,0xc420458100, 0x3d, 0xfce03bc, 0xc420458140, 0x32, 0xc420458180, 0x37, ...)
        /home/ops/specs-bot/src/hook/hook.go:37 +0x13b fp=0xc420506ed8 sp=0xc420505ed0 pc=0x78ad1b
main.main.func1(0x9aa500, 0xc420292000, 0xc420412000)
        /home/ops/specs-bot/src/specs-bot/main.go:81 +0x4d9 fp=0xc42050bcf0 sp=0xc420506ed8 pc=0x78d6f9
net/http.HandlerFunc.ServeHTTP(0xc420099980, 0x9aa500, 0xc420292000, 0xc420412000)
        /root/go/src/net/http/server.go:1947 +0x44 fp=0xc42050bd18 sp=0xc42050bcf0 pc=0x6870a4
net/http.(*ServeMux).ServeHTTP(0xde8c00, 0x9aa500, 0xc420292000, 0xc420412000)
        /root/go/src/net/http/server.go:2337 +0x130 fp=0xc42050bd58 sp=0xc42050bd18pc=0x688d10
net/http.serverHandler.ServeHTTP(0xc42009b110, 0x9aa500, 0xc420292000, 0xc420412000)
        /root/go/src/net/http/server.go:2694 +0xbc fp=0xc42050bd88 sp=0xc42050bd58 pc=0x689d4c
net/http.(*conn).serve(0xc4204c83c0, 0x9aaa00, 0xc42027a140)
        /root/go/src/net/http/server.go:1830 +0x651 fp=0xc42050bfc8 sp=0xc42050bd88pc=0x6860c1
runtime.goexit()
        /root/go/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc42050bfd0 sp=0xc42050bfc8 pc=0x4605f1
created by net/http.(*Server).Serve
        /root/go/src/net/http/server.go:2795 +0x27b

goroutine 1 [IO wait]:
internal/poll.runtime_pollWait(0x7fd3b2a1ff00, 0x72, 0x0)
        /root/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc42013e298, 0x72, 0xc4200a8300, 0x0, 0x0)
        /root/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc42013e298, 0xffffffffffffff00, 0x0, 0x0)
        /root/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Accept(0xc42013e280, 0x0, 0x0, 0x0,0x0, 0x0, 0x0, 0x0)
        /root/go/src/internal/poll/fd_unix.go:372 +0x1a8
net.(*netFD).accept(0xc42013e280, 0xc4204c8440, 0xc420053ca0, 0x40c2d8)
        /root/go/src/net/fd_unix.go:238 +0x42
net.(*TCPListener).accept(0xc4200a40d8, 0xc420053cd0, 0x40b237, 0xc4204c8440)
        /root/go/src/net/tcpsock_posix.go:136 +0x2e
net.(*TCPListener).AcceptTCP(0xc4200a40d8, 0xc420053d18, 0xc420053d20, 0x18)
        /root/go/src/net/tcpsock.go:246 +0x49
net/http.tcpKeepAliveListener.Accept(0xc4200a40d8, 0x958c50, 0xc4204c83c0, 0x9aaac0, 0xc420099bc0)
        /root/go/src/net/http/server.go:3216 +0x2f
net/http.(*Server).Serve(0xc42009b110, 0x9aa6c0, 0xc4200a40d8, 0x0, 0x0)
        /root/go/src/net/http/server.go:2770 +0x1a5
net/http.(*Server).ListenAndServe(0xc42009b110, 0xc42009b110, 0x13)
        /root/go/src/net/http/server.go:2711 +0xa9
net/http.ListenAndServe(0xc4200a0770, 0xe, 0x0, 0x0, 0xc4200b45a0, 0x0)
        /root/go/src/net/http/server.go:2969 +0x7a
main.main()
        /home/ops/specs-bot/src/specs-bot/main.go:104 +0x606

goroutine 19 [select, 12385minutes]:
database/sql.(*DB).connectionOpener(0xc4200b6dc0, 0x9aaa00, 0xc4200a8440)
        /root/go/src/database/sql/sql.go:935 +0x119
created by database/sql.OpenDB
        /root/go/src/database/sql/sql.go:634 +0x178

goroutine 20 [select, 12385minutes]:
database/sql.(*DB).connectionResetter(0xc4200b6dc0, 0x9aaa00, 0xc4200a8440)
        /root/go/src/database/sql/sql.go:948 +0x12a
created by database/sql.OpenDB
        /root/go/src/database/sql/sql.go:635 +0x1ae

goroutine 408 [runnable]:
mutex.Lock(0x79a)
        /home/ops/specs-bot/src/mutex/mutex.go:19 +0x9c
hook.HandlePullRequestPayload(0xc4205580c8, 0x7, 0x79a,0xc4205d8180, 0x3d, 0xfce03bc, 0xc4205d81c0, 0x32, 0xc4205d8280, 0x37, ...)
        /home/ops/specs-bot/src/hook/hook.go:37 +0x13b
main.main.func1(0x9aa500, 0xc420432000, 0xc420180200)
        /home/ops/specs-bot/src/specs-bot/main.go:81 +0x4d9
net/http.HandlerFunc.ServeHTTP(0xc420099980, 0x9aa500, 0xc420432000, 0xc420180200)
        /root/go/src/net/http/server.go:1947 +0x44
net/http.(*ServeMux).ServeHTTP(0xde8c00, 0x9aa500, 0xc420432000, 0xc420180200)
        /root/go/src/net/http/server.go:2337 +0x130
net/http.serverHandler.ServeHTTP(0xc42009b110, 0x9aa500, 0xc420432000, 0xc420180200)
        /root/go/src/net/http/server.go:2694 +0xbc
net/http.(*conn).serve(0xc420566000, 0x9aaa00, 0xc4200a80c0)
        /root/go/src/net/http/server.go:1830 +0x651
created by net/http.(*Server).Serve
        /root/go/src/net/http/server.go:2795 +0x27b

goroutine 409 [IO wait]:
internal/poll.runtime_pollWait(0x7fd3b2a1fc90, 0x72, 0xc420272658)
        /root/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc42013e018, 0x72, 0xffffffffffffff00, 0x9a7760, 0xdbb4d0)
        /root/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc42013e018, 0xc4201fa000, 0x1, 0x1)
        /root/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc42013e000, 0xc4201fa0d1, 0x1, 0x1, 0x0, 0x0, 0x0)
        /root/go/src/internal/poll/fd_unix.go:157 +0x17d
net.(*netFD).Read(0xc42013e000, 0xc4201fa0d1, 0x1, 0x1,0xc4202727a0, 0x0, 0xc42008a060)
        /root/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc42000e010, 0xc4201fa0d1, 0x1, 0x1, 0x0, 0x0, 0x0)
        /root/go/src/net/net.go:176 +0x6a
net/http.(*connReader).backgroundRead(0xc4201fa0c0)
        /root/go/src/net/http/server.go:668 +0x5a
created by net/http.(*connReader).startBackgroundRead
        /root/go/src/net/http/server.go:664 +0xce

goroutine 390 [IO wait]:
internal/poll.runtime_pollWait(0x7fd3b2a1fe30, 0x72, 0xc420269658)
        /root/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc4205ae018, 0x72, 0xffffffffffffff00, 0x9a7760, 0xdbb4d0)
        /root/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc4205ae018, 0xc4204e0100, 0x1, 0x1)
        /root/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc4205ae000, 0xc4204e0101, 0x1, 0x1, 0x0, 0x0, 0x0)
        /root/go/src/internal/poll/fd_unix.go:157 +0x17d
net.(*netFD).Read(0xc4205ae000, 0xc4204e0101, 0x1, 0x1,0x8aa120, 0x8c7ee0, 0x64c600)
        /root/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc420260000, 0xc4204e0101, 0x1, 0x1, 0x0, 0x0, 0x0)
        /root/go/src/net/net.go:176 +0x6a
net/http.(*connReader).backgroundRead(0xc4204e00f0)
        /root/go/src/net/http/server.go:668 +0x5a
```

### PR DESCRIPTION
**Pull request by vabd**
Fixes #7

### CODE REVIEW

File: src/mutex/mutex.go [hunk 1]
```diff
@@ -4,11 +4,17 @@ import (
...
 
 // Lock locks the mutex for a given proposal, after instanciating if it doesn't
 // exist.
 func Lock(number int64) {
+	// Avoid multiple goroutines using the map at the same time.
+	lock.Lock()
+
 	// Lock the existing mutex if there's one.
 	if m, exists := mutexes[number]; exists {
 		m.Lock()
```
Comment (GordonF42) [inline src/mutex/mutex.go:21]: You never unlock `lock` if `mutexes[number]` exists

File: src/mutex/mutex.go [hunk 2]
```diff
@@ -4,11 +4,17 @@ import (
 	"sync"
 )
 
-var mutexes = make(map[int64]*sync.Mutex)
+var (
+	mutexes = make(map[int64]*sync.Mutex)
+	lock    = new(sync.Mutex)
```
Comment (GordonF42) [inline src/mutex/mutex.go]: I'd rather name it `mapLock` or `mapMutex`

**Review by GordonF42 — DISMISSED** [review]

File: src/mutex/mutex.go (hunk 1 above)
Comment (vabd) [reply src/mutex/mutex.go:21]: Addressed by deferring call to Unlock() rather than having it happen at the end of the function.

**Review by GordonF42 — COMMENTED** [review]
LGTM

**Pull request merged**